packages/plugins/src/hardhat/simulate.ts

```diff
@@ -161,9 +161,28 @@ export const simulate = async (
 
   if (code !== 0) {
     const networkName = getNetworkNameForChainId(BigInt(chainId))
-    throw new Error(
-      `Simulation failed for ${networkName} at block number ${block.number}. Reason:\n${stderr}`
-    )
+    let errorMessage: string = `Simulation failed for ${networkName} at block number ${block.number}.`
+    try {
+      // Attempt to decode the error message. This try-statement could theoretically throw an error
+      // if `stdout` isn't a valid JSON string.
+      const error = JSON.parse(stdout)
+
+      // If the stack trace includes the error message, we only use the stack trace so that we don't
+      // display the error reason twice.
+      if (
+        typeof error.stack === 'string' &&
+        error.stack.includes(error.message)
+      ) {
+        errorMessage += `\n\n${error.stack}`
+      } else {
+        // Display both the error message and the stack trace.
+        errorMessage += `\n\n${error.message}\n\n${error.stack}`
+      }
+    } catch {
+      // Do nothing if an error occurs while attempting to decode the error from the child process.
+      // We'll throw a generic error shortly.
+    }
+    throw new Error(errorMessage)
```

Comment:
I would include the raw value of stdout in the error message. It's hard to predict what the value of stdout would be in this case, so I'd advise airing on the side of caution by passing it through to the user in case it's useful.

Reply:
Good idea, added that